### PR TITLE
chore(http_mock): improvements & tapping

### PR DESCRIPTION
**Pull request by StarlightIbuki**
### Summary

Introduce tapping for testing, especially OIDC's behavior when sending requests to IDP.

bugfix: assert truthy
alias for not_all
more checks
better comment&document

### Checklist

- [ ] Cherry-pick to EE master

### Issue reference

N/A

### CODE REVIEW

File: spec/02-integration/01-helpers/03-http_mock_spec.lua [hunk 1]
```diff
@@ -220,3 +221,47 @@ describe("http_mock config", function()
     assert(pl_file.access_time(pid_filename) ~= nil, "mocking not in the correct place")
   end)
 end)
+
+describe("#http_mock.tapping", function()
```
Comment (Tieske) [inline spec/02-integration/01-helpers/03-http_mock_spec.lua:225]: is the hashtag a left over?; `#`

File: spec/02-integration/01-helpers/03-http_mock_spec.lua [hunk 2]
```diff
@@ -220,3 +221,47 @@ describe("http_mock config", function()
...
+  local tapped, tapped_port
+  lazy_setup(function()
+    tapped, tapped_port = http_mock.new()
+    tapped:start()
+  end)
+  lazy_teardown(function()
+    tapped:stop(true)
+  end)
+
+  it("it works", function()
```
Comment (Tieske) [inline spec/02-integration/01-helpers/03-http_mock_spec.lua:235]: the idea of the test framework is to make it easy to read; to the `it(` should not be repeated. The description will be added to the describe description; eg. `http_mock.tapping it works`

```suggestion
  it("works", function()
```

would create a test description `http_mock.tapping works`, which is how the framework was intended

**Review by ADD-SP — APPROVED** [review]
LGTM

File: spec/helpers/http_mock.lua [hunk 3]
```diff
@@ -43,11 +44,12 @@ local function default_field(tbl, key, default)
 end
 
 -- create a mock instance which represents a HTTP mocking server
--- @param listens: the listen directive of the mock server, defaults to "0.0.0.0:8000"
--- @param code: the code of the mock server, defaults to a simple response.
--- @param opts: options for the mock server, left it empty to use the defaults
--- @return: a mock instance
--- @usage
+---@param listens table|string|number|nil: the listen directive of the mock server, defaults to a random available port
```
Comment (Tieske) [inline spec/helpers/http_mock.lua:47]: what's up with these doc-comments? they should be `ldoc` based, but this is a different format.

The old format was correct; `-- @param ...`

**Review by Tieske — CHANGES_REQUESTED** [review]

**Comment by Tieske** [comment]
please ensure that the documentation properly renders;

here's how to check:
```shell
cd kong/spec
ldoc .
open docs/index.html
```

It requires to have `ldoc` installed; `luarocks install ldoc`

File: spec/02-integration/01-helpers/03-http_mock_spec.lua (hunk 1 above)
Comment (StarlightIbuki) [reply spec/02-integration/01-helpers/03-http_mock_spec.lua]: Yes. Removed.

File: spec/helpers/http_mock.lua (hunk 3 above)
Comment (StarlightIbuki) [reply spec/helpers/http_mock.lua]: Got a little bit confused as I can find both formats in our code base. The `---@param` works with my VSC plugin's hover-over hint so I chose this.
Changing to the correct format.

**Review by oowl — APPROVED** [review]

**Comment by Tieske** [comment]
I pushed a commit with some docs fixes, please verify, and if possible extend to the other modules that are currently undocumented. These new get rendered properly.

File: spec/helpers/http_mock/tapping.lua [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+  }
+end
+
+--- create a new http_mock.tapping instance with a tapping route.
+-- @tparam string|number target the target host/port of the tapping route
+-- @tparam[opt] table|string|number listens the listen directive of the mock server, defaults to a random available port
+-- @tparam[opt="servroot_tapping"] string prefix the prefix of the mock server
+-- @tparam[opt] table log_opts see `http_mock.new`
+-- @treturn http_mock.tapping a tapping instance
+-- @treturn number the port the mock server listens to
```
Comment (Tieske) [inline spec/helpers/http_mock/tapping.lua:34]: this 2nd return value is wrong, seems a copy-paste error, please update that it is correct.

Comment (Tieske) [reply spec/helpers/http_mock/tapping.lua:34]: hmmm seems it only has one return value, so please remove the second one.

Comment (StarlightIbuki) [reply spec/helpers/http_mock/tapping.lua:34]: its returning the function return value and lua will have 2 return value in this case.

Comment (StarlightIbuki) [reply spec/helpers/http_mock/tapping.lua:34]: (the called function returns 2 so the calling function have a list of 2 values to return)

**Comment by Tieske** [comment]
fyi; the reason I'm pushing on this is because we have been neglecting the documentation in the helpers. Mostly because not everyone realizes this is actually user-facing stuff. We should pay more attention to this.

Plugin developers will use `pongo docs` to [render and show](https://github.com/Kong/kong-pongo/blob/master/README.md?plain=1#L58) the documentation when developing their plugins.

**Comment by Tieske** [comment]
Added some more fixes, but `ldoc` simply sucks at multi-file modules/classes 😞 

anyway, since I now committed to this PR, I can no longer approve it. Feel free to merge. @oowl @ADD-SP

**Review by Tieske — APPROVED** [review]

**Comment by Tieske** [comment]
still could approve 🤷 , so just did

**Pull request merged**